File: src/variantfoldrna/workflow/source/snpfold.go
// Go implementation of the SNPfold algorithm
// Original python code was obtained from https://github.com/Halvee/SNPfold

package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"math/rand"
	"bufio"
	"github.com/montanaflynn/stats"
	"time"
	// "sync"
	// "testing"
)

func main() {

	// Read in the command line inputs 
	seq := flag.String("seq", "", "Sequence")
	mutation := flag.String("mut", "", "Mutation")
	temp := flag.String("T", "", "Temperature")
	flag.Parse()

	// Check to make sure that we have a value for each of the inputs
	if *seq == "" || *mutation == "" || *temp == "" {
		flag.PrintDefaults()
		os.Exit(1)
	}

	// Check to make sure there are no N's in the sequence 
	if strings.Contains(*seq, "N") {
		fmt.Println("N is in the sequence")
		os.Exit(1)
	}

	// Get the reference allele 
	ref := string((*mutation)[0])

	// Get the position of the mutation
	pos, err := strconv.Atoi((*mutation)[1:len(*mutation)-1])

	// Check to see if there was an error when converting the mutation
	// position to an integer
	if err != nil {
		fmt.Println("Error converting position to integer:", err)
		os.Exit(1)
	}

	if string((*seq)[pos-1]) != ref {
		fmt.Println("Reference allele does not match sequence at position", pos)
		os.Exit(1)
	}

	if strings.Contains(*mutation, "N") {
		fmt.Println("N is in the mutation")
		os.Exit(1)
	}

	// Get the base-pairing probability matrix, the 
	// freeEnergy, and the ensembleDiversity for the REF and ALT alleles
	bppMatrix, freeEnergy, ensembleDiversity, nil := getBPPMatrix(*seq, *temp)
	bppMatrixMut, freeEnergyMut, ensembleDiversityMut, nil := getBPPMatrix(getMutatedSequence(*seq, *mutation), *temp)

	colSums := getSumMatrixCols(bppMatrix)
	colSumsMut := getSumMatrixCols(bppMatrixMut)

	pearson, err := stats.Pearson(colSums, colSumsMut)
	if err != nil {
		fmt.Println("Error calculating Pearson correlation coefficient:", err)
		os.Exit(1)
	}

	fmt.Println(fmt.Sprintf("%s,%s,%s,%s,%f", freeEnergy, ensembleDiversity, freeEnergyMut, ensembleDiversityMut, pearson))
}


func getBPPMatrix(sequence string, temp string) ([][]float64, string, string, error) {
	// Create a temporary FASTA file with the given sequence
	tempFasta, name, err := createTempFasta(sequence)
	if err != nil {
		return nil, "", "", err
	}
	defer os.Remove(tempFasta) // Remove the temporary FASTA file when done

	// Execute RNAfold command
	output, err := exec.Command("RNAfold", "-p", "--noPS", "-i", tempFasta, "-T", temp).Output()
	if err != nil {
		return nil, "", "", fmt.Errorf("error running RNAfold: %w", err)
	}

	// Extract free energy of the ensemble and ensemble diversity
	freeEnergy, ensembleDiversity := extractEnsembleInfo(string(output))
	// if err != nil {
	// 	return nil, "", "", fmt.Errorf("error extracting ensemble info: %w", err)
	// }

	// Parse output to extract base pair probability matrix
	bppMatrix  := parseOutput(".sq" + name + "_dp.ps", sequence)
	// if err != nil {
	// 	return nil, "", "", fmt.Errorf("error parsing output: %w", err)
	// }

	return bppMatrix, freeEnergy, ensembleDiversity, nil
}

func extractEnsembleInfo(outputStr string) (string, string) {

	// Split the output by newline characters
	lines := strings.Split(outputStr, "\n")

	// Extract free energy of the ensemble
	freeEnergyLine := lines[2]
	freeEnergy := strings.Split(freeEnergyLine, " ") // Extract the part after '('
	freeEnergyLength := len(freeEnergy)
	freeEnergyEle := freeEnergy[freeEnergyLength-1] // Extract the last element
	freeEnergyEle = strings.TrimRight(freeEnergyEle, ") ")    // Remove trailing ')' and space
	freeEnergyEle = strings.TrimLeft(freeEnergyEle, "( ") 
	// Extract ensemble diversity
	ensembleDiversityLine := lines[5]
	ensembleDiversity := strings.Split(ensembleDiversityLine, ";")[1]
	ensembleDiversity = strings.Split(ensembleDiversityLine, " ")[10] // Extract the part after ';'
	ensembleDiversity = strings.TrimSpace(ensembleDiversity)

	return freeEnergyEle, ensembleDiversity

}


func parseOutput(file_name string, sequence string) ([][]float64){

	bppMatrix := make([][]float64, len(sequence))
	for i := range bppMatrix {
		bppMatrix[i] = make([]float64, len(sequence))
	}

	file, _ := os.Open(file_name)

	defer file.Close()

	// Create a scanner to read the file line by line
	scanner := bufio.NewScanner(file)

	// Iterate over each line in the file
	for scanner.Scan() {
		line := scanner.Text() // Get the current line

		if strings.Contains(line, "ubox") {
			parts := strings.Fields(line)
			if len(parts) == 4 && parts[3] == "ubox" {
				i, _ := strconv.Atoi(parts[0])
				j, _ := strconv.Atoi(parts[1])
				val, _ := strconv.ParseFloat(parts[2], 64)
				bppMatrix[i-1][j-1] = val * val
				bppMatrix[j-1][i-1] = val * val
			}
		}
	}
	os.Remove(file_name)
	return bppMatrix 
}



func createTempFasta(sequence string) (string, string, error) {
	tempName := getRandomName()
	tempFasta, err := ioutil.TempFile("", "sequence_" + tempName)
	if err != nil {
		return "", "", fmt.Errorf("error creating temporary file: %w", err)
	}
	defer tempFasta.Close()
	
	// Placeholder for generating a unique name
	_, err = tempFasta.WriteString(fmt.Sprintf(">.sq%s\n%s", tempName, sequence))
	if err != nil {
		return "", "", fmt.Errorf("error writing to temporary file: %w", err)
	}
	return tempFasta.Name(), tempName, nil
}


func getRandomName() string {
	const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	result := make([]byte, 12)
	rand.Seed(time.Now().UnixNano())
	for i := range result {
		result[i] = chars[rand.Intn(len(chars))]
	}
	return string(result)
}

func getMutatedSequence(sequence string, mutation string) string {
	index, _ := strconv.Atoi(mutation[1 : len(mutation)-1])

	mutatedSequence := sequence[:index-1] + string(mutation[len(mutation)-1]) + sequence[index:]
	return mutatedSequence
}

func getSumMatrixCols(matrix [][]float64) []float64 {
	colSums := make([]float64, len(matrix))
	for i := range matrix {
		for j := range matrix {
			colSums[j] += matrix[i][j]
		}
	}
	return colSums
}

func roundFloat(val float64, precision uint) float64 {
    ratio := math.Pow(10, float64(precision))
    return math.Round(val*ratio) / ratio
}

File: src/variantfoldrna/workflow/scripts/snpfold.py
##################################################################
# This script will calculate the SNPfold score for a given SNP
#
# Original code was obtained from https://github.com/Halvee/SNPfold
# Updated for python3 by Kobie Kirven
# Assmann and Bevilacqua Labs
##################################################################

# Imports
import argparse
import numpy as np
import subprocess
import tempfile
import os
import string
import random


parser = argparse.ArgumentParser(description="Calculate SNPfold score for a given SNP")
parser.add_argument("-seq", dest="seq", help="Sequence")
parser.add_argument("-mut", dest="mutation", help="Mutation")
parser.add_argument("-T", dest="temp", help="Temperature")
args = parser.parse_args()

# Check to make sure "N" is not in the sequence
if "N" in args.seq:
    raise ValueError("N is in the sequence")

# Check to make sure "N" is not in the mutation
if "N" in args.mutation:
    raise ValueError("N is in the mutation")


def get_temp_fasta(sequence):
    """
    Writes a temporary fasta file with the given sequence
    """
    fn = tempfile.NamedTemporaryFile(delete=False)
    name = "".join(
        random.choice(string.ascii_uppercase + string.digits) for _ in range(6)
    )
    fn.write(b">.sq" + name.encode("utf-8") + b"\n")
    fn.write(sequence.encode("utf-8"))
    return fn.name, name


def get_bppm(sequence):
    """
    Runs RNAfold on the given sequence and returns the base pair probibility matrix
    """
    temp_fasta, name = get_temp_fasta(sequence)
    output = subprocess.check_output(
        ["RNAfold", "-p", "--noPS", "-i", temp_fasta, "-T", args.temp]
    ).decode("utf-8")

    # output = output.split("\n")
        # Split the output string into lines
    lines = output.split("\n")

    # Extract free energy of the ensemble
    freeEnergyLine = lines[2]
    freeEnergy = freeEnergyLine.split(" ")[-1]  # Extract the part after '('
    freeEnergyEle = freeEnergy.split("(")[-1].strip(")").strip(" ")  # Extract the last element and remove trailing ')' and space
    
    # Extract ensemble diversity
    ensembleDiversityLine = [x for x in lines if "ensemble diversity" in x][0]
    ensembleDiversity = ensembleDiversityLine.split(";")[1]
    ensembleDiversity = ensembleDiversity.split(" ")[3]  # Extract the part after ';'
    ensembleDiversity = ensembleDiversity.strip()  # Remove leading and trailing whitespace


    bpp_matrix = np.zeros((len(sequence), len(sequence)), dtype=float)

    with open(".sq" + name + "_dp.ps") as fn:
        lines = fn.readlines()
        for line in lines:
            if "ubox\n" in line:
                line = line.split(" ")
                if len(line) == 4:
                    if line[3] == "ubox\n":
                        bpp_matrix[int(line[0]) - 1, int(line[1]) - 1] = float(
                            line[2]
                        ) * float(line[2])
                        bpp_matrix[int(line[1]) - 1, int(line[0]) - 1] = float(
                            line[2]
                        ) * float(line[2])
    os.remove(temp_fasta)
    os.remove(".sq" + name + "_dp.ps")

    return bpp_matrix, freeEnergyEle, ensembleDiversity


def get_sum_matrix_cols(bpp_matrix):
    "Get the column sums from a BPP matrix"
    return np.sum(bpp_matrix, axis=0)


def get_pearson(list1, list2):
    """
    Calculate the Pearson correlation coefficient between two lists
    """
    return round(np.corrcoef(list1, list2)[0][1], 3)


if __name__ == "__main__":
    # Get the BPP matrices
    bpp_matrix, ref_dg, ref_ed = get_bppm(args.seq)
    bpp_matrix_mut, alt_dg, alt_ed = get_bppm(
        args.seq[: int(args.mutation[1:-1]) - 1]
        + args.mutation[-1]
        + args.seq[int(args.mutation[1:-1]) :]
    )

    # Get the column sums
    col_sums = get_sum_matrix_cols(bpp_matrix)
    col_sums_mut = get_sum_matrix_cols(bpp_matrix_mut)

    # Calculate the Pearson correlation coefficient
    pearson = get_pearson(col_sums, col_sums_mut)
    print(f"{ref_dg},{ref_ed},{alt_dg},{alt_ed},{pearson}")
